daq/runner.py

```diff
@@ -814,6 +816,9 @@ def _create_gateway(self, device):
             gateway.set_tap_intf(self.network.tap_intf)
 
         try:
+            if device:
+                device.gateway = gateway
+                self.network.register_device(device)
```

Comment:
I don't think the "register_device" name is appropriate here. This function is really for setting the network up for gateway (which should be device independent). I think rename to self.network.allocate_gateway_ports (or something like that) and pass in the port_set is all that's needed here. Then you can revert line 688, 794-795 and 819-821.